packages/runner/src/runner.ts

```diff
@@ -475,4 +474,346 @@ export class Runner {
+      });
+    }
+    if (
+      BigInt(inputCkbCapacitySum) ===
+      BigInt(ckbWithdrawalCapacity) +
+        BigInt(outputCkbCapacitySumForSudtCustodianCells)
+    ) {
+      // so lucky with it, just return
+      return txSkeleton;
+    }
+    // build ckb change custodian cell
+    const custodianLock: Script = {
+      code_hash: this._deploymentConfig().custodian_lock.code_hash,
+      hash_type: this._deploymentConfig().custodian_lock.hash_type,
+      args: DEFAULT_CUSTODIAN_LOCK_ARGS,
+    };
+    let ckbChangeCustodian: Cell = {
+      cell_output: {
+        lock: custodianLock,
+        type: undefined,
+        capacity: "0x0",
+      },
+      data: utils.toBigUInt128LE(BigInt(0)),
+    };
+    const minimalCapacity = minimalCellCapacity(ckbChangeCustodian);
+    const changeCapacity =
+      BigInt(inputCkbCapacitySum) -
+      BigInt(ckbWithdrawalCapacity) -
+      BigInt(outputCkbCapacitySumForSudtCustodianCells);
+    if (BigInt(changeCapacity) >= BigInt(minimalCapacity)) {
+      // With enough input ckb capacity to build a minimal change cell
+      ckbChangeCustodian.cell_output.capacity =
+        "0x" + changeCapacity.toString(16);
+      txSkeleton = txSkeleton.update("outputs", (outputs) => {
+        return outputs.push(ckbChangeCustodian);
+      });
+      return txSkeleton;
+    }
+    // Gonna collect some more ckb to build the change cell
+    for (const cell of validCustodianCells) {
+      if (!cell.cell_output.type) {
+        const key = getInputKey(cell);
+        if (previousInputs.has(key)) {
+          continue;
+        }
+        previousInputs.add(key);
+        const inputCkbCapacity = BigInt(cell.cell_output.capacity);
+        inputCkbCapacitySum += inputCkbCapacity;
+        txSkeleton = txSkeleton.update("inputs", (inputs) => {
+          return inputs.push(cell);
+        });
+        if (
+          BigInt(inputCkbCapacitySum) >=
+          BigInt(ckbWithdrawalCapacity) +
+            BigInt(outputCkbCapacitySumForSudtCustodianCells) +
+            BigInt(minimalCapacity)
+        ) {
+          break;
+        }
+      }
+    }
+    // TODO collect ckb from other sudt custodian cells if all non sudt custodian cells' ckb capacity is insufficient,
+    // but so far just throw an error here.
+    if (
+      BigInt(inputCkbCapacitySum) <
+      BigInt(ckbWithdrawalCapacity) +
+        BigInt(outputCkbCapacitySumForSudtCustodianCells) +
+        BigInt(minimalCapacity)
+    ) {
+      throw new Error("Insufficient CKB capacity in valid custodian cells");
+    }
+    const newChangeCapacity =
+      BigInt(inputCkbCapacitySum) -
+      BigInt(ckbWithdrawalCapacity) -
+      BigInt(outputCkbCapacitySumForSudtCustodianCells);
+    ckbChangeCustodian.cell_output.capacity =
+      "0x" + newChangeCapacity.toString(16);
+    txSkeleton = txSkeleton.update("outputs", (outputs) => {
+      return outputs.push(ckbChangeCustodian);
+    });
+    return txSkeleton;
+  }
+
+  _buildWithdrawalLockArgs(
```

Comment:
Please write a denormalizer for RawWithdrawalRequest, and a normalizer for WithdrawalLockArgs.

Reply:
updated.